ESQL: URL encoding changes

mouhc1ne opened the pull request:
- Change `URL_ENCODE` to encode spaces as `+`.
- Add a new `URL_ENCODE_COMPONENT` scalar function, which encodes spaces as `%20`.
- Both encoding functions encode all characters in the input except for alphanumerics, `.`, `-`, `_`, and `~`.
- Allow `URL_DECODE` to fail gracefully if the input can't be decoded, by returning null and adding a warning in the header, similar to other scalar functions.
- Update csv-tests:
  - Test decoding for both `+` and `%20` back to space.
  - Reduce the length of doc lines to less than 76 chars, as mentioned in the contribution guide.
- Minor changes to the documentation.

mouhc1ne committed:
ESQL: URL encoding changes
- Change URL_ENCODE to encode spaces as +.
- Add a new URL_ENCODE_COMPONENT scalar function, which encodes spaces as %20.
- Both encoding functions encode all characters in the input except for alphanumerics, ., -, _, and ~.
- Allow URL_DECODE to fail gracefully if the input can't be decoded, by returning null and adding a warning in the header, similar to other scalar functions.
- Update csv-tests:
  - Test decoding for both + and %20 back to space.
  - Reduce the length of doc lines to less than 76 chars, as mentioned in the contribution guide.
- Minor changes to the documentation.

mouhc1ne commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractUrlEncodeDecodeTestCase.java:
```diff
@@ -87,19 +87,21 @@ public static TestCaseSupplier.TestCase createTestCaseWithRandomString(
-        return URLEncoder.encode(plain, StandardCharsets.UTF_8);
+    private static String encode(boolean useUrlComponentEncoder, String plain) {
+        byte[] plainBytes = plain.getBytes(StandardCharsets.UTF_8);
+        byte[] encoded = useUrlComponentEncoder ? UrlCodecUtils.encodeUrl(plainBytes) : UrlCodecUtils.encodeUrlComponent(plainBytes);
```
Nailed it. Also adding a space char somewhere in the randomized tests to make them fail if this changes.

nik9000 commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/UrlCodecUtils.java:
```diff
@@ -0,0 +1,92 @@
+
+        try {
+            encoded = urlEncodeCodec.encode(bytes);
+        } catch (EncoderException ex) {
```
Honestly, reading `PercentCodec` I think you are better off making your own that doesn't throw weird exception and can operate on `BytesRef` and can deal with a `BreakingBytesRefBuilder` for the result directly. You can certainly have a test that throws random input into both of them and compares them. But I think you've written almost as much code to work around the interface mismatch as you would to actually do the work. And if you do the work yourself it's easier to optimize it if we have to.

mouhc1ne commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/UrlCodecUtils.java:
```diff
@@ -0,0 +1,92 @@
+        urlEncodeCodec = new PercentCodec(b1, true);
+
+        // encodes spaces as '%20'
+        byte[] b2 = buildUnsafeBytes(Collections.emptySet());
```
Done.

mouhc1ne commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/UrlDecodeTests.java:
Did you mean unit tests specifically? There are csv tests that cover null/invalid inputs.

ivancea replied:
Yes, unit tests here. We have some test duplication between the different suites, but better safe than sorry IMO

mouhc1ne commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/UrlEncode.java:
```diff
@@ -83,9 +86,10 @@ public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvalua
-        String input = val.utf8ToString();
-        String encoded = URLEncoder.encode(input, StandardCharsets.UTF_8);
-        return new BytesRef(encoded);
+        byte[] input = new byte[val.length];
```
Changed to return the passed `BytesRef` if the input doesn't change at all after encoding. Otherwise, we use the scratch buffer to allocate just enough space for the encoded bytes.

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/UrlEncode.java:
```diff
@@ -45,11 +47,15 @@ private UrlEncode(StreamInput in) throws IOException {
         returnType = "keyword",
         preview = true,
-        description = "URL encodes the input.",
+        description = "URL-encodes the input. All characters are percent-encoded except for alphanumerics, "
+            + "`.`, `-`, `_`, and `~`. Spaces are encoded as `+`.",
```
I was reading usages of this function in other techs, and I'm not sure about this. Sorry, as I didn't see/understand this before.
The `*Component*` function is quite clear, or at least should be: It encodes following the RFC ([RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986) I think?). Every reserved character, no questions asked.

The `URL_Encode` however, as defined in [JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI) and [.NET](https://learn.microsoft.com/en-us/dotnet/api/system.web.httputility.urlencode?view=net-8.0) (As some examples), shouldn't touch characters that are part of the URL syntax. Meaning that it's quite error-prone, but only to be used as a best-effort to fix broken URLs or visualization purposes (I guess). But I couldn't find any specific standard about this function. From what I understand, it "just" ignores the chars used in an URL, and that's it.

What made me look deeper into this, was that the difference between both functions here is just the "+", which was getting weirder the more I thought about it

mouhc1ne committed:
ESQL: URL encoding changes
- Manually perform percent-encoding directly on the BreakingBytesRefBuilder scratch area, and only if the input was gonna change after encoding.
- Use Apache's PercentEncode as the ground truth during unit testing.
- Add test cases with fixed strings.

mouhc1ne committed:
ESQL: URL encoding changes
- Add fixed test cases that fail decoding
- Randomize the encoder used in decoding tests
- Add url_encode_component capability to some encode/decode tests whose behavior changed to ensure they only run if the url_encode_component capability is available
